Extend STM32 QSPI with QER (SFDP, DTS), custom quad write opcode, and 1-1-4 read mode

GeorgeCGV opened the pull request:
Extends existing STM32 QSPI NOR driver.

Adds:

* Support every QE requirement - can be specified in DTS. However, SFDP BFP DW15 bypasses the DTS value (if DW15 is present).
* Add support to specify what quad program (_write_) opcode is going to be used:

  - `PP_1_1_4` (**_new_**)
  - `PP_1_4_4` (_used originally_)

  Some flash ICs support a different set of opcodes for programming and reading. The inspiration is taken from the 
  `nrf_qspi_nor` driver, but `spi_nor` defines are used. 
* Add support for 1-1-4 read mode. The driver auto-detects the fastest supported quad read mode from SFDP BFP 

This PR enables STM32 QSPI NOR driver usage with different kinds of QSPI Flash ICs.

Tested using custom H750 board with W25Q128JVSIQ.

<details>
  <summary>Flash test output</summary>

> [00:00:00.000,000] <inf> flash_stm32_qspi: flash_stm32_qspi_init: W25Q128JVSIQ: SFDP v 1.5 AP ff with 1 PH
> [00:00:00.000,000] <inf> flash_stm32_qspi: flash_stm32_qspi_init: PH0: ff00 rev 1.5: 16 DW @ 80
> [00:00:00.000,000] <inf> flash_stm32_qspi: W25Q128JVSIQ: 16 MiBy flash
> [00:00:00.000,000] <inf> flash_stm32_qspi: Quad read mode 4 instr [0x6b] supported
> [00:00:00.000,000] <inf> flash_stm32_qspi: Quad read mode 7 instr [0xeb] supported
> [00:00:00.000,000] <inf> flash_stm32_qspi: Quad read mode 7 instr [0xeb] will be used
> [00:00:00.000,000] <inf> flash_stm32_qspi: QE Requirement mode: 4
> [00:00:00.000,000] <inf> flash_stm32_qspi: Quad mode enabled
>
> uart:~$ flash test W25Q128JVSIQ 0x0 0x1000 3
> Erase OK.
> Write OK.
> Erase OK.
> Write OK.
> Erase OK.
> Write OK.
> Erase-Write test done.
>
> uart:~$ flash erase W25Q128JVSIQ 0x0 0x1000
> Erase success.
>
> uart:~$ flash read  W25Q128JVSIQ 0x0 0x16
> 00000000: ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff |........ ........|
> 00000010: ff ff ff ff ff ff                                |......           |
>
> uart:~$ flash write W25Q128JVSIQ 0x0 0xAABBCCDD
> Write OK.
> Verified.
>
> uart:~$ flash read  W25Q128JVSIQ 0x0 0x16
> 00000000: dd cc bb aa ff ff ff ff  ff ff ff ff ff ff ff ff |........ ........|
> 00000010: ff ff ff ff ff ff                                |......           |
>
> uart:~$

</details

GeorgeCGV committed:
drivers: flash: spi_nor: remove not used QE bit define

Removes unused QE bit define.

Signed-off-by: Georgij Cernysiov <geo.cgv@gmail.com>

GeorgeCGV committed:
drivers: falsh: spi_nor: add PP_1_1_4 and rename 4PP

* Renames 4PP define to PP_1_4_4
* Adds PP_1_1_4 define

That matches linux kernel defines a bit more.

Signed-off-by: Georgij Cernysiov <geo.cgv@gmail.com>

GeorgeCGV committed:
drivers: flash: spi_nor: add SR3 read and write defines

Adds Status Register 3 read and write opcode defines.

Signed-off-by: Georgij Cernysiov <geo.cgv@gmail.com>

GeorgeCGV committed:
drivers: flash: jesd216: correct comment mistake

Corrects type name in the comment for the DW15 QER field.

Signed-off-by: Georgij Cernysiov <geo.cgv@gmail.com>

GeorgeCGV committed:
dts: bindings: flash_controller: stm32-qspi-nor: add writeoc property

Adds writeoc property to allow quad write opcode selection.

Signed-off-by: Georgij Cernysiov <geo.cgv@gmail.com>

GeorgeCGV committed:
drivers: flash: stm32: qspi: minor cleanup

Removes double line and log output regarding
device initialization.

Signed-off-by: Georgij Cernysiov <geo.cgv@gmail.com>

GeorgeCGV committed:
drivers: flash: stm32: qspi: support DTS writeoc

Adds support for DTS writeoc.

Uses 1-4-4 mode by default (as the original driver).

Signed-off-by: Georgij Cernysiov <geo.cgv@gmail.com>

GeorgeCGV committed:
drivers: flash: stm32: qspi: support 1-4-4 and 1-1-4 quad read modes

Adds support for 1-4-4 and 1-1-4 read modes.

SFDP is used to query for available read instructions, then the
fastest one is used.

Signed-off-by: Georgij Cernysiov <geo.cgv@gmail.com>

erwango commented on dts/bindings/flash_controller/st,stm32-qspi-nor.yaml:
```diff
@@ -49,3 +49,18 @@ properties:
+        - "PP_1_4_4"      # Quad data line SPI, PP 1-4-4 (0x38)
+      description: |
+        Specify the number of data lines and opcode used for writing.
+        If not provided pp will be selected.
```
Duplicate vs last line ?

GeorgeCGV replied:
Reworded with additional details.

erwango commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -132,38 +137,57 @@ static inline void qspi_set_address_size(const struct device *dev,
+		switch (cmd->Instruction) {
+		case SPI_NOR_CMD_PP_1_1_4:
+			cmd->AddressMode = QSPI_ADDRESS_1_LINE;
+			cmd->DataMode = QSPI_DATA_4_LINES;
+			cmd->DummyCycles = 0;
```
``
			cmd->DataMode = QSPI_DATA_4_LINES;
			cmd->DummyCycles = 0;
``
Why not factorizing these lines ?

erwango commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -29,19 +30,21 @@
 
 #define STM32_QSPI_RESET_GPIO DT_INST_NODE_HAS_PROP(0, reset_gpios)
 #if STM32_QSPI_RESET_GPIO
-#include <drivers/gpio.h>
+#include <zephyr/drivers/gpio.h>
 #endif
```
While you're at it, I'd find it reasonnable to remove the conditional around this header inclusion

erwango commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -132,38 +137,57 @@ static inline void qspi_set_address_size(const struct device *dev,
+		switch (dev_data->mode) {
+		case JESD216_MODE_114:
+			cmd->AddressMode = QSPI_ADDRESS_1_LINE;
+			cmd->DataMode = QSPI_DATA_4_LINES;
```
Why not factorizing this line ?

erwango commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -1173,6 +1317,16 @@ static struct flash_stm32_qspi_data flash_stm32_qspi_dev_data = {
+			(JESD216_DW15_QER_VAL_S1B6)),
+	.qspi_write_cmd = COND_CODE_1(DT_INST_NODE_HAS_PROP(0, writeoc),
+			(_CONCAT(SPI_NOR_CMD_,
+				 DT_STRING_UPPER_TOKEN(DT_DRV_INST(0),
```
```suggestion
				 DT_STRING_TOKEN(DT_DRV_INST(0),
```

erwango replied:
Also please introduce macros to increase readability

GeorgeCGV replied:
Wasn't sure about the macro names. I hope they are unique enough as they relate to DT and certain properties.

erwango commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -875,34 +992,63 @@ static int spi_nor_process_bfp(const struct device *dev,
+		/* bypass DTS QE requirement (if possible) */
+		rc = jesd216_bfp_decode_dw15(php, bfp, &dw15);
+		if (rc < 0) {
+			LOG_ERR("Unable to decode QE requirement [DW15]: %d", rc);
```
Are we able to continue in this case ?

GeorgeCGV replied:
Yes, as `qer_type` can be defined in the DTS - `quad_enable_requirements` and if not, then `JESD216_DW15_QER_VAL_S1B6` is used to stay functionally equal to the driver that had no `quad_enable_requirements` support (_it was hardcoded to `JESD216_DW15_QER_VAL_S1B6`_).

erwango commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -133,17 +136,29 @@ static inline void qspi_set_address_size(const struct device *dev,
+			cmd->AddressMode = QSPI_ADDRESS_4_LINES;
+			break;
+		default:
+			return -ENOTSUP;
```
Is that possible at this step ? If not, what about a mere if/else

GeorgeCGV replied:
It is not possible to have a not supported value here as well as in the reading prepare function.

However, that helps to clearly communicate to another dev or a maintainer that nothing else is supported. This might help to detect errors when DTS `writeoc` enum is extended or when somebody adds other modes (potentially refactors to support dual/octal modes).

erwango commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -939,13 +1027,18 @@ static int spi_nor_process_bfp(const struct device *dev,
 
 		LOG_INF("Quad read mode %d instr [0x%x] will be used", data->mode, res.instr);
 
+
```
Superflous line.

GeorgeCGV committed:
drivers: flash: stm32: qspi: move prepare cmd is enable check

Moves IS_ENABLE check from prepare program and read to the caller.

Signed-off-by: Georgij Cernysiov <geo.cgv@gmail.com>

GeorgeCGV committed:
drivers: flash: stm32: qspi: support DTS quad enable requirements

Adds support for DTS quad_enable_requirements property.

Signed-off-by: Georgij Cernysiov <geo.cgv@gmail.com>

GeorgeCGV committed:
drivers: flash: stm32: qspi: support SFDP quad enable requirements

Parses SFDP QER related DW15 to use as the source for QER.

Signed-off-by: Georgij Cernysiov <geo.cgv@gmail.com>

GeorgeCGV committed:
drivers: flash: stm32: qspi: always include gpio header

Always include gpio header.

Signed-off-by: Georgij Cernysiov <geo.cgv@gmail.com>

GeorgeCGV commented:
- Rebased
- https://github.com/zephyrproject-rtos/zephyr/pull/45433#discussion_r866847422
  - Added extra comments in the code to reduce confusion
  - Changed `LOG_ERR` to `LOG_WRN`

FRASTM commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -1108,8 +1107,6 @@ static int flash_stm32_qspi_init(const struct device *dev)
 #endif /* CONFIG_FLASH_PAGE_LAYOUT */
 
-	LOG_INF("Device %s initialized", dev->name);
-
```
any reason why the LOG_INF removal ?

GeorgeCGV replied:
Not present in the majority of the drivers. Maybe this can be done as `DBG` if required.

FRASTM commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -132,38 +136,55 @@ static inline void qspi_set_address_size(const struct device *dev,
 	struct flash_stm32_qspi_data *dev_data = dev->data;
 
-	if (IS_ENABLED(STM32_QSPI_USE_QUAD_IO) && dev_data->flag_quad_io_en) {
+	if (dev_data->flag_quad_io_en) {
```
```suggestion
	
	cmd->AddressMode = ((dev_data->mode == JESD216_MODE_114)
				? QSPI_ADDRESS_1_LINE
				: QSPI_ADDRESS_4_LINES);
```

GeorgeCGV replied:
`switch` allows us to detect the case when some mode is not yet supported (i.e. `JESD216_MODE_444`) and return `-ENOTSUP` in such case. That eases figuring out why something is not working.

Alternatively, I can add `__ASSERT` then the ternary operator is ok.

FRASTM commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -323,7 +344,13 @@ static int flash_stm32_qspi_read(const struct device *dev, off_t addr,
 	qspi_set_address_size(dev, &cmd);
-	qspi_prepare_quad_read(dev, &cmd);
+	if (IS_ENABLED(STM32_QSPI_USE_QUAD_IO)) {
+		ret = qspi_prepare_quad_read(dev, &cmd);
```
the test of  the STM32_QSPI_USE_QUAD_IO is done out of the function
and then the `qspi_prepare_quad_read() `function itself checks the `dev_data->flag_quad_io_en`
compare to previoulsy "if (IS_ENABLED(STM32_QSPI_USE_QUAD_IO) && dev_data->flag_quad_io_en) "
Correct ?

GeorgeCGV replied:
Yes, the `IS_ENABLED` is moved outside.

What do you think about removing `flag_quad_io_en`? I don't think there is a need for it. 
If the driver is in Quad mode (`STM32_QSPI_USE_QUAD_IO` is enabled), then the `init` will fail if QE enabling fails... So there won't be a case that QE enabling failed and we still use the QSPI in Quad mode...

FRASTM replied:
I agree

GeorgeCGV committed:
drivers: flash: stm32: qspi: remove flag_quad_io_en field

There is no need in the `flag_quad_io_en` field.
When QE enabling failed, then the driver `init` fails as well.

There is no way to use qspi in quad mode if qe is not enabled.

Signed-off-by: Georgij Cernysiov <geo.cgv@gmail.com>

GeorgeCGV committed:
drivers: flash: stm32: qspi: use ternary and asserts in quad prepare

Use ternary operation and asserts to reduce the footprint
and code size.

Signed-off-by: Georgij Cernysiov <geo.cgv@gmail.com>

GeorgeCGV commented:
Rebased, got rid of `flag_quad_io_en` and switched to the ternary operator and asserts in the quad command preparation.

erwango reviewed (approved):
A bit of commit number inflation, but I won't be picky this close to the DV freeze.

mbolivar-nordic commented:
@nvlsianpu I'm going to merge this to make it in time for code freeze since the changes to the common SPI NOR header seem quite harmless. If I messed up, I apologize, and we can revert.

GeorgeCGV commented:
> @nvlsianpu I'm going to merge this to make it in time for code freeze since the changes to the common SPI NOR header seem quite harmless. If I messed up, I apologize, and we can revert.

- `SPI_NOR_QE_BIT` - wasn't used by any existing driver
- `SPI_NOR_CMD_4PP` - was used by the STM32 QSPI driver. Therefore, renaming shall be fine
- `RDSR3`/`WRSR3` - new
- `SPI_NOR_CMD_PP_1_4_4` - new
